# ============================================================
# Notebook 02 : Data Cleaning
# Author : Sarvan Bishnoi
# ============================================================


In [ ]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)

# Load Dataset
df = pd.read_csv("/content/hotel_bookings.csv.zip")

print(df.shape)

# Working Copy
df_clean = df.copy()

(119390, 32)


In [ ]:
duplicate_count = df_clean.duplicated().sum()

print("Duplicate Rows :", duplicate_count)

df_clean.drop_duplicates(inplace=True)

print(df_clean.shape)

Duplicate Rows : 31994
(87396, 32)


In [ ]:
missing = pd.DataFrame({
    "Missing Values": df_clean.isnull().sum(),
    "Percentage": round(df_clean.isnull().mean()*100,2)
})

missing = missing.sort_values(
    by="Percentage",
    ascending=False
)

missing

,Missing Values,Percentage
company,82137,93.98
agent,12193,13.95
country,452,0.52
hotel,0,0.00
arrival_date_month,0,0.00
arrival_date_week_number,0,0.00
lead_time,0,0.00
is_canceled,0,0.00
stays_in_weekend_nights,0,0.00
stays_in_week_nights,0,0.00


In [ ]:
# children -> median
df_clean["children"] = df_clean["children"].fillna(
    df_clean["children"].median()
)

# country -> Unknown
df_clean["country"] = df_clean["country"].fillna("Unknown")

# agent -> 0
df_clean["agent"] = df_clean["agent"].fillna(0)

# company -> 0
df_clean["company"] = df_clean["company"].fillna(0)

In [ ]:
df_clean.isnull().sum()

,0
hotel,0
is_canceled,0
lead_time,0
arrival_date_year,0
arrival_date_month,0
arrival_date_week_number,0
arrival_date_day_of_month,0
stays_in_weekend_nights,0
stays_in_week_nights,0
adults,0


In [ ]:
df_clean["reservation_status_date"] = pd.to_datetime(
    df_clean["reservation_status_date"]
)

In [ ]:
# Total Guests

df_clean["total_guests"] = (
    df_clean["adults"]
    + df_clean["children"]
    + df_clean["babies"]
)

# Total Nights

df_clean["total_nights"] = (
    df_clean["stays_in_weekend_nights"]
    + df_clean["stays_in_week_nights"]
)

In [ ]:
print(df_clean.shape)

print(df_clean.info())

(87396, 34)
<class 'pandas.core.frame.DataFrame'>
Index: 87396 entries, 0 to 119389
Data columns (total 34 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   hotel                           87396 non-null  object        
 1   is_canceled                     87396 non-null  int64         
 2   lead_time                       87396 non-null  int64         
 3   arrival_date_year               87396 non-null  int64         
 4   arrival_date_month              87396 non-null  object        
 5   arrival_date_week_number        87396 non-null  int64         
 6   arrival_date_day_of_month       87396 non-null  int64         
 7   stays_in_weekend_nights         87396 non-null  int64         
 8   stays_in_week_nights            87396 non-null  int64         
 9   adults                          87396 non-null  int64         
 10  children                        87396 non-null  float64       

In [ ]:
df_clean.to_csv(
    "cleaned_hotel_bookings.csv",
    index=False
)

print("Dataset Saved Successfully ✅")

Dataset Saved Successfully ✅


In [ ]:
# children -> median
df_clean["children"] = df_clean["children"].fillna(
    df_clean["children"].median()
)

# country -> Unknown
df_clean["country"] = df_clean["country"].fillna("Unknown")

# agent -> 0 (No travel agent)
df_clean["agent"] = df_clean["agent"].fillna(0)

# company -> 0 (No company booking)
df_clean["company"] = df_clean["company"].fillna(0)

In [ ]:
missing_after = pd.DataFrame({
    "Missing Values": df_clean.isnull().sum(),
    "Percentage": round(df_clean.isnull().mean()*100,2)
})

missing_after[missing_after["Missing Values"] > 0]

,Missing Values,Percentage


Empty DataFrame
Columns: [Missing Values, Percentage]
Index: []

In [ ]:
df_clean["reservation_status_date"] = pd.to_datetime(
    df_clean["reservation_status_date"]
)

In [ ]:
# Total Guests

df_clean["total_guests"] = (
    df_clean["adults"]
    + df_clean["children"]
    + df_clean["babies"]
)

# Total Nights

df_clean["total_nights"] = (
    df_clean["stays_in_weekend_nights"]
    + df_clean["stays_in_week_nights"]
)

In [ ]:
print(df_clean.shape)

df_clean.head()

(87396, 34)


,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,children,babies,meal,country,market_segment,distribution_channel,is_repeated_guest,previous_cancellations,previous_bookings_not_canceled,reserved_room_type,assigned_room_type,booking_changes,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,total_guests,total_nights
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,0.0,0,BB,PRT,Direct,Direct,0,0,0,C,C,3,No Deposit,0.0,0.0,0,Transient,0.0,0,0,Check-Out,2015-07-01,2.0,0
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,0.0,0,BB,PRT,Direct,Direct,0,0,0,C,C,4,No Deposit,0.0,0.0,0,Transient,0.0,0,0,Check-Out,2015-07-01,2.0,0
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,0.0,0,BB,GBR,Direct,Direct,0,0,0,A,C,0,No Deposit,0.0,0.0,0,Transient,75.0,0,0,Check-Out,2015-07-02,1.0,1
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,0.0,0,BB,GBR,Corporate,Corporate,0,0,0,A,A,0,No Deposit,304.0,0.0,0,Transient,75.0,0,0,Check-Out,2015-07-02,1.0,1
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,0.0,0,BB,GBR,Online TA,TA/TO,0,0,0,A,A,0,No Deposit,240.0,0.0,0,Transient,98.0,0,1,Check-Out,2015-07-03,2.0,2


In [ ]:
df_clean[[
    "adults",
    "children",
    "babies",
    "total_guests",
    "stays_in_weekend_nights",
    "stays_in_week_nights",
    "total_nights"
]].head()

,adults,children,babies,total_guests,stays_in_weekend_nights,stays_in_week_nights,total_nights
0,2,0.0,0,2.0,0,0,0
1,2,0.0,0,2.0,0,0,0
2,1,0.0,0,1.0,0,1,1
3,1,0.0,0,1.0,0,1,1
4,2,0.0,0,2.0,0,2,2


In [ ]:
df_clean.to_csv(
    "cleaned_hotel_bookings.csv",
    index=False
)

print("✅ Cleaned dataset saved successfully.")

✅ Cleaned dataset saved successfully.


# Conclusion

### Data Cleaning Summary

- Removed 31,994 exact duplicate records.
- Filled missing values in children using median.
- Replaced missing country values with "Unknown".
- Replaced missing agent and company values with 0.
- Converted reservation_status_date into datetime format.
- Created two new features:
  - total_guests
  - total_nights

### Final Dataset

- Rows : 87,396
- Columns : 34

The dataset is now clean and ready for Exploratory Data Analysis (EDA).